### PR TITLE
Add rekor client method to upload a record

### PR DESCRIPTION
**Pull request by loosebazooka**
- Also adds a cert generator to bypass fulcio/dex in testing

### CODE REVIEW

File: src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.nio.charset.StandardCharsets;
+import java.security.cert.Certificate;
+import java.util.Base64;
+import java.util.HashMap;
+import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
+import org.bouncycastle.util.encoders.Hex;
+
+public class HashedRekordRequest {
+
+  private final Hashedrekord hashedrekord;
```
Comment (patflynn) [inline src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java:31]: Is the (lackof) camecase a typo here or unavoidable code gen artifact?

File: src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  private HashedRekordRequest(Hashedrekord hashedrekord) {
+    this.hashedrekord = hashedrekord;
+  }
+
+  /**
+   * Create a new HashedRekorRequest.
+   *
+   * @param artifactDigest the sha256 digest of the artifact (not hex/base64 encoded)
+   * @param leafCert the leaf certificate used to verify {@code signature}, usually obtained from
```
Comment (patflynn) [inline src/main/java/dev/sigstore/rekor/client/HashedRekordRequest.java:41]: for custom key signed artifacts (not using fulcio), is there a leaf cert to submit, or do you just use the singer's public key? It's all still a little muddled for me so asking out of curiosity,

File: src/test/java/dev/sigstore/rekor/client/RekorClientTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    var signer = Signers.newEcdsaSigner();
+    var signature = signer.sign(data.getBytes(StandardCharsets.UTF_8));
+
+    // create a fake signing cert (not fulcio/dex)
+    var cert = CertGenerator.newCert(signer.getPublicKey());
+
+    var req = HashedRekordRequest.newHashedRekordRequest(artifactDigest, cert, signature);
+
+    // this tests directly against rekor in staging, it's a bit hard to bring up a rekor instance
+    // without docker compose.
```
Comment (patflynn) [inline src/test/java/dev/sigstore/rekor/client/RekorClientTest.java:57]: todo? I think Jason was working on doing this for his maven testing.

File: src/test/java/dev/sigstore/testing/CertGenerator.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.operator.ContentSigner;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+
+/**
+ * A certificate generator, useful when trying to talk to rekor without actually using fulcio/oidc.
+ */
+public class CertGenerator {
```
Comment (patflynn) [inline src/test/java/dev/sigstore/testing/CertGenerator.java:47]: ah so would this be used for the custom key flow too?

**Review by patflynn — APPROVED** [review]

File: src/test/java/dev/sigstore/rekor/client/RekorClientTest.java (hunk 3 above)
Comment (loosebazooka) [reply src/test/java/dev/sigstore/rekor/client/RekorClientTest.java:57]: I don't know what the right solution is. I think sigstore-scaffolding is fine, but it's super heavy weight.

Issue: https://github.com/sigstore/sigstore-java/issues/27

File: src/test/java/dev/sigstore/testing/CertGenerator.java (hunk 4 above)
Comment (loosebazooka) [reply src/test/java/dev/sigstore/testing/CertGenerator.java:47]: if you use a custom key, you probably don't want a certificate -- you just use a pem encoded public key?

**Pull request merged**